LinkedList implementation of TimeRangeGroup

talvarez0507 opened the pull request:
This is the linked list implementation of TimeRangeGroup. It uses a linear search for hasTimeRange and edits the list in place for adding and deleting. The tests are exactly the same as the tests for the ArrayList implementation.

talvarez0507 committed:
create linked list implementation

talvarez0507 committed:
add tests for linked list implementation

These are the exact same tests that Holly wrote, but modified for the linked list implementation.

talvarez0507 committed:
improve time complexity of add method

The add method had a quadratic time complexity and I fixed it so that it is linear now.

talvarez0507 committed:
implement ListIterator

talvarez0507 committed:
remove unnecessary comments

talvarez0507 commented on portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java:
```diff
@@ -0,0 +1,557 @@
+package com.google.sps.data;
```
These are exactly the same tests as ArrayListTimeRangeGroupTest.java which is already in master except this tests LinkedListTimeRangeGroup instead.

talvarez0507 committed:
nit binary -> linear

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+
+/**
+ * Models an implementation of the TimeRangeGroup model using LinkedList. This is a slightly
+ * modified version of what hollyyuqizheng wrote for the ArrayList implementation.
```
I don't think you need to mention hollyyuqizheng. You can write that it modified ArrayListTime...

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+  // This list of all the time ranges will be sorted by start time ascending
+  // in add and delete methods. The time ranges stored in this list are
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
```
I noticed that this class and ArrayListTimeRangeGroup share a fair amount of code. I would suggest that you use an abstract class to avoid duplicating the code.

ex:
abstract class AbstractListTimeRangeGroup implements TimeRangeGroup {
  protected List<TimeRange> allTimeRanges;

 ...
}

public class LinkedListTimeRangeGroup extends AbstractListTimeRangeGroup implements TimeRangeGroup {
  ...
}

talvarez0507 replied:
Done. I will change the ArrayList implementation to extend AbstractListTimeRangeGroup in another PR.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        iterator.remove();
+        continue;
```
Consider the case where the initial internal time range list just has 1pm-2pm, and timeRange is 12pm-3pm. Then I think this line gets triggered, and the internal list stays at 1pm-2pm, rather than being updated to 12pm-3pm.

talvarez0507 replied:
Fixed. I had to rewrite a bit of the code but all of these corner cases having to due with a single element or adding one element all the way at the end work fine now. Tests were also added for these cases.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+        iterator.add(lastExaminedTimeRange);
+        iterator.next();
+        lastExaminedTimeRange = currentRange;
+      }
```
Should we consider the else case? It can get triggered for example if the internal list was just 1pm-2pm, and timeRange is 3pm-4pm.

talvarez0507 replied:
The else case means that the currentRange and lastExaminedTimeRange do not overlap and that lastExaminedTimeRange is strictly later than currentRange. On some further iteration the else if will trigger and insert it into the list once it comes before something in the list so we don't have to worry about the else case.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
```
This method can be a helper method defined in the abstract class.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> getAllTimeRangesIterator() {
```
The abstract class can have this method.

rgcase replied:
nit: It's named `.iterator()` in the other implementation and you should update yours to match, so it can be used in a foreach loop.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+      if (currentRange.overlaps(timeRangeToDelete)) {
```
It might be simpler to continue in the opposite case, so you don't have to indent the inner block.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+   * original list, deleting [3:30 - 5:30] will result in two new ranges: [3 - 3:30] and [5:30 - 6].
+   */
+  public void deleteTimeRange(TimeRange timeRangeToDelete) {
+
```
Unnecessary line break

talvarez0507 replied:
Removed.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+   * the target range is contained within this current range.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
+
```
Unnecessary line break

talvarez0507 replied:
Removed.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java:
```diff
@@ -0,0 +1,557 @@
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class LinkedListTimeRangeGroupTest {
```
Consider using a parameterized test, which runs the same test suite on LinkedListTimeRangeGroup as well as ArrayListTimeRangeGroup.

Reference:
https://junit.org/junit4/javadoc/4.12/org/junit/runners/Parameterized.html

Maybe we can name it as TimeRangeGroupTest

talvarez0507 replied:
Done.

rgcase commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,187 @@
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
```
nit: You can just do `timeRanges.forEach(addTimeRange)`.

talvarez0507 replied:
I was not able to make this work. I got the error message "cannot find symbol addTimeRange" even though it is a method in the class.

talvarez0507 committed:
create AbstractListTimeRangeGroup.java

talvarez0507 committed:
create ArrayListTimeRangeGroup.java

talvarez0507 committed:
update to extend AbstractListTimeRangeGroup

talvarez0507 committed:
create parameterized tests

talvarez0507 committed:
delete individual test

talvarez0507 committed:
remove typo

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java:
```diff
@@ -0,0 +1,43 @@
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  public static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
```
I think this method can be protected instead of public.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> iterator() {
```
(Optional) You can add the @Override annotation here (capital O, which somehow is not capitalized when I submit the comment on GitHub).

Reference:
https://docs.oracle.com/javase/tutorial/java/annotations/predefined.html

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java:
```diff
@@ -0,0 +1,146 @@
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
```
You can consider adding the override annotation

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java:
```diff
@@ -0,0 +1,146 @@
+   */
+  public ArrayListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new ArrayList<TimeRange>();
+    timeRanges.forEach(
```
Consider pulling out this forEach into a helper method. I think it can be a protected method declared in the abstract class. Then here you can invoke it.

(As a result, you may have to declare addTimeRange in the abstract method.)

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java:
```diff
@@ -0,0 +1,669 @@
+    return Arrays.asList(ArrayListTimeRangeGroup.class, LinkedListTimeRangeGroup.class);
+  }
+
+  private Class timeRangeGroupClass;
```
I think you can do,

@Parameter Class timeRangeGroupClass;

Then you won't need the TimeRangeGroupTest constructor below.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,157 @@
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
```
As I suggested earlier, I think you can just invoke the helper method here for forEach.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,157 @@
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)
+          && !lastExaminedTimeRange.equals(currentRange)) {
```
Do we need the ! equals check?

talvarez0507 replied:
If we remove the !equals then when we try to add an event that already exists it is removed and never added.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:
```diff
@@ -0,0 +1,157 @@
+      // The case for when two time ranges need to be merged.
+      // Similar to the case above, it will eventually be replaced so we can
+      // remove currentRange
+      if (lastExaminedTimeRange.overlaps(currentRange)
```
Can this be an else if? If so, then we won't need the ! contains check.

talvarez0507 replied:
I think this can't be an else if because the if statement that comes before this and the else if statement after this are not mutually exclusive.

jamesnam-google reviewed (approved):
Approving the change with comments

hollyyuqizheng commented on portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java:
```diff
@@ -0,0 +1,669 @@
+    // Result:       |-----------|
+    Instant timeRangeOneStart = Instant.now();
+    Instant timeRangeOneEnd = timeRangeOneStart.plusSeconds(3000);
+    Instant timeRangeTwoStart = timeRangeOneStart.plusSeconds(500);
```
timeRangeTwoStart should probably be `timeRangeOneStart.minusSeconds(500)` ?

talvarez0507 replied:
Fixed.

hollyyuqizheng commented on portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java:
```diff
@@ -0,0 +1,669 @@
+    }
+    List<TimeRange> expected = Arrays.asList(timeRangeOne);
+
+    timeRangeGroup.addTimeRange(timeRangeOne);
```
Hmmm what if you try to construct a new TimeRange instance called `timeRangeTwo`, and initialize this new time range with the exact start and end times of timeRangeOne? I think in this case the test passes for array list time range group, at least it did when I tried it on my own.

talvarez0507 replied:
Seems like I just had outdated code, ArrayListTimeRangeGroup passes this test just fine now.

talvarez0507 committed:
fix AbstractListTimeRangeGroup.java

I made a new helper method for adding all the ranges and changed the iterator() method to protected.

talvarez0507 committed:
use new helper method

talvarez0507 committed:
add override annotations

talvarez0507 committed:
use new helper method

talvarez0507 committed:
use @parameter instead of constructor

talvarez0507 committed:
Merge branch 'master' into TimeRangeGroupLinkedList
